### PR TITLE
Custom response

### PR DESCRIPTION
**Pull request by ghost**
@signalplus-dev

### CODE REVIEW

File: app/assets/javascripts/components/forms/calendar.jsx [hunk 1]
```diff
@@ -1,25 +1,34 @@
...
+      valid,
+      meta: { touched, error} ,
+      ...props,
+    } = this.props;
 
-  handleChange(e) {
-    const key = 'expirationDate'
-    this.setState({date: e});
-  }
+    const inputValue = moment(input.value) || moment();
```
Comment (signalplus-dev) [inline app/assets/javascripts/components/forms/calendar.jsx]: Can you make this 1 week away? Doesn't make sense to make it expire the same that you make the custom response.

**Comment by ghost** [comment]
TODO Items here:

Spec for listen signals controller
Validation for Timed Responses

File: app/assets/javascripts/components/forms/validations.jsx [hunk 2]
```diff
@@ -13,20 +14,48 @@ export function signalNameValidator(name) {
...
-  } else if (repeatResponse.length > 140) {
-    return 'Your tweet response cannot exceed 140 characters.';
+export function expirationDateValidator(date) {
+  if (!date) {
+    return 'Expiration date must be set on a custom response';
+  }
+
+  const selectedDate = moment(date, 'YYYY-MM-DD');
+  if (selectedDate < moment()) {
+    return 'Expiration date cannot be in the past';
```
Comment (signalplus-dev) [inline app/assets/javascripts/components/forms/validations.jsx:59]: What about custom responses that are supposed to be expired? Do you only want to do this for new ones?

File: app/assets/stylesheets/components/calendar.scss [hunk 3]
```diff
@@ -1,5 +1,17 @@
...
+  width: 130px !important;
+  margin-top: 10px;
+  border-radius: 10px;
+  width: 150px;
+  font-size: 16px;
+  font-weight: bold;
+  font-style: normal;
+  font-stretch: normal;
+  border-color: $button-blue;
+  border-color: var(--windows-blue);
```
Comment (signalplus-dev) [inline app/assets/stylesheets/components/calendar.scss]: ?

File: app/controllers/api/v1/listen_signals_controller.rb [hunk 4]
```diff
@@ -96,16 +100,30 @@ def listen_signal_template_types
...
+  end
+
   def put_update
     signal_params = update_signal_params
-    @listen_signal.update!(signal_params)
+    response_params = create_response_params
 
-    update_response(@listen_signal.default_response, params[:default_response])
-    update_response(@listen_signal.repeat_response, params[:repeat_response])
+    ActiveRecord::Base.transaction do
```
Comment (signalplus-dev) [inline app/controllers/api/v1/listen_signals_controller.rb]: This controller is getting bloated. Can we move the create and update listen signal logic to a service?

File: app/controllers/api/v1/listen_signals_controller.rb [hunk 5]
```diff
@@ -116,10 +134,38 @@ def create_response(message, type, response_group)
     Response.create_response(message, type, response_group)
   end
 
-  def create_grouped_response(default_response_msg, repeat_response_msg)
-    response_group = create_response_group
-    create_response(default_response_msg, Response::Type::DEFAULT, response_group)
-    create_response(repeat_response_msg, Response::Type::REPEAT, response_group)
+  def create_timed_response(message, expiration_date, response_group)
+    Response.create_timed_response(message, Response::Type::TIMED, expiration_date, response_group)
```
Comment (signalplus-dev) [inline app/controllers/api/v1/listen_signals_controller.rb]: If the method is called `.create_timed_response`, it seems superfluous to pass in the type of response. It seems like it should be hard coded within the method itself.

File: app/models/response_group.rb [hunk 6]
```diff
@@ -50,4 +50,9 @@ def repeat_response
   def expired_response
     responses.find(&:expired?)
   end
+
+  # @return [Response]
```
Comment (signalplus-dev) [inline app/models/response_group.rb]:
```
# @return [Array<Response>]
```

**Comment by ghost** [comment]
@signalplus-dev  Yo rick, moved update/create into `listen_signal_handler` service

File: spec/services/listen_signal/listen_signal_handler_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+          'default_response': 'default test ',
+          'repeat_response': 'repeat test',
+          'responses': [
+            { message: 'timed1 test', expiration_date: '2016-12-25' }
+          ]
+        }
+      }
+
+      it 'deletes a timed response' do
+        expect(listen_signal.timed_responses.count).to eq(1)
```
Comment (signalplus-dev) [inline spec/services/listen_signal/listen_signal_handler_spec.rb:101]: You should really be testing that the original timed responses are now soft deleted.

File: spec/services/listen_signal/listen_signal_handler_spec.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    it 'does not create response group' do
+      expect(ResponseGroup.count).to eq(1)
+    end
+
+    it 'updates listen signal' do
+      expect(listen_signal.name).not_to eq(name)
+    end
+
+    it 'updates default response' do
+      expect(listen_signal.default_response.message).not_to eq(default_response.message)
```
Comment (signalplus-dev) [inline spec/services/listen_signal/listen_signal_handler_spec.rb:77]: A good way to test these is by
```ruby
expect {
  subject.update
}.to change {
  listen_signal.reload.default_response.message
}.from(default_response.message).to('default test')
```

**Review by signalplus-dev — APPROVED** [review]

**Pull request merged**